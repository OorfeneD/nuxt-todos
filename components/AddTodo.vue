<template>
  <div class="flex w-full justify-between align-middle">
    <input v-model="title" class="focus:shadow-inner transition shadow-xs focus:outline-none border border-gray-200 text-gray-600 px-3 py-2 rounded-md flex-grow mr-2" type="text" placeholder="todo title ...">
    <button class="hover:bg-indigo-500 text-gray-600 hover:text-white transition px-2 py-2 border border-gray-200 rounded-md hover:shadow-sm" @click="$emit('addtodo', { content: { title }})">
      Add
    </button>
  </div>
</template>

<script>
export default {
  data () {
    return {
      title: ''
    }
  }
}
</script>

<style>

</style>
