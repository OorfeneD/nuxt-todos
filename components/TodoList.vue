<template>
  <ul class="py-1">
    <todo
      v-for="todo in todos"
      :key="todo.id"
      :todo="todo"
      @removetodo="$emit('removetodo', $event)"
      @updatetodo="$emit('updatetodo', $event)"
    />
    <li v-if="todos.length < 1" class="text-sm text-gray-400 text-center py-2">
      There is nothing here. Add your first ToDo
    </li>
  </ul>
</template>

<script>
import { mapGetters } from 'vuex'
import Todo from './Todo.vue'
export default {
  components: { Todo },
  computed: {
    ...mapGetters(['todos'])
  }
}
</script>

<style>

</style>
