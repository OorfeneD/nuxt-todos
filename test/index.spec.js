import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueX from 'vuex'
import { todo, badTodo } from './mocks/todo.mock'
import index from '@/pages/index'
import AddTodo from '~/components/AddTodo.vue'
import TodoList from '~/components/TodoList.vue'

const localVue = createLocalVue()
localVue.use(VueX)
global.alert = jest.fn()
describe('index page', () => {
  let actions
  let getters
  let store
  let emptyGetters
  let emptyStore
  const spies = {
    showError: jest.fn()
  }
  const preInit = () => {
    actions = {
      addTodo: jest.fn(),
      removeTodo: jest.fn(),
      updateTodo: jest.fn()
    }
    getters = {
      isLoggedIn: () => true
    }
    emptyGetters = {
      isLoggedIn: () => false
    }
    // eslint-disable-next-line
    store = new VueX.Store({
      actions,
      modules: {
        user: {
          getters,
          namespaced: true
        }
      }
    })
    // eslint-disable-next-line
    emptyStore = new VueX.Store({
      actions,
      modules: {
        user: {
          getters: emptyGetters,
          namespaced: true
        }
      }
    })
    spies.showError = jest.spyOn(index.methods, 'showError')
  }
  describe('render', () => {
    beforeAll(preInit)
    test('is Vue instance', () => {
      const wrapper = shallowMount(index, { store, localVue })
      expect(wrapper.vm).toBeTruthy()
    })
    test('is match snapshot with logged in user', () => {
      const wrapper = shallowMount(index, { store, localVue })
      expect(wrapper.html()).toMatchSnapshot()
    })
    test('is match snapshot with logged out user', () => {
      const wrapper = shallowMount(index, { store: emptyStore, localVue })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
  describe('vuex', () => {
    let wrapper
    beforeAll(() => {
      preInit()
      wrapper = shallowMount(index, {
        store,
        localVue
      })
    })
    test('dispatches `addTodo` action on `AddTodo` events', () => {
      wrapper.findComponent(AddTodo).vm.$emit('addtodo', todo)
      expect(actions.addTodo).toHaveBeenCalled()
    })
    test('`_addTodo` method calls `showError` on invalid data', () => {
      wrapper.findComponent(AddTodo).vm.$emit('addtodo', badTodo)
      expect(spies.showError).toHaveBeenCalled()
    })
    test('calls `showError` action on `AddTodo` events', () => {
      wrapper.findComponent(AddTodo).vm.$emit('error', { msg: '' })
      expect(spies.showError).toHaveBeenCalled()
      wrapper.vm.$forceUpdate()
    })
    test('dispatches `removeTodo` action on `TodoList` component events', () => {
      wrapper.findComponent(TodoList).vm.$emit('removetodo', { todo })
      expect(actions.removeTodo).toHaveBeenCalled()
    })
    test('dispatches `updateTodo` action on `TodoList` component events', () => {
      wrapper.findComponent(TodoList).vm.$emit('updatetodo', { newStatus: true, todo })
      expect(actions.updateTodo).toHaveBeenCalled()
    })
  })
})
