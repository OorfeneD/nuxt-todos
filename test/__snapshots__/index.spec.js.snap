// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index page render is match snapshot with logged in user 1`] = `
"<div class=\\"w-screen flex flex-col items-center py-2\\">
  <div class=\\"py-3 px-2\\">
    <h1 class=\\"w-full text-center text-2xl\\">
      ToDo List
    </h1> <span class=\\"text-gray-400 text-sm\\">
      by <a href=\\"https://github.com/OorfeneD\\" title=\\"GitHub OorfeneD\\" class=\\"text-blue-300 hover:bg-gray-100 hover:text-blue-500 transition p-1 rounded\\"><icon-stub github=\\"true\\"></icon-stub> OorfeneD</a></span>
  </div>
  <div class=\\"w-full sm:w-96 rounded-0 sm:rounded-xl border border-gray-100 shadow-md p-5\\">
    <add-todo-stub></add-todo-stub>
    <todo-list-stub></todo-list-stub>
  </div>
</div>"
`;

exports[`index page render is match snapshot with logged out user 1`] = `
"<div class=\\"w-screen flex flex-col items-center py-2\\">
  <div class=\\"py-3 px-2\\">
    <h1 class=\\"w-full text-center text-2xl\\">
      ToDo List
    </h1> <span class=\\"text-gray-400 text-sm\\">
      by <a href=\\"https://github.com/OorfeneD\\" title=\\"GitHub OorfeneD\\" class=\\"text-blue-300 hover:bg-gray-100 hover:text-blue-500 transition p-1 rounded\\"><icon-stub github=\\"true\\"></icon-stub> OorfeneD</a></span>
  </div>
  <div>
    Authenticate to work with TODOs
  </div>
</div>"
`;
